src/badguy/tarantula.cpp

```diff
@@ -0,0 +1,386 @@
+    m_physic.set_velocity_x(MOVE_SPEED);
+
+  m_state = STATE_APPROACHING;
+  return APPROACH;
+}
+
+bool
+Tarantula::try_drop()
+{
+  using RaycastResult = CollisionSystem::RaycastResult;
+
+  // Assuming the player has already been checked...
+
+  Vector eye(get_bbox().get_middle().x, get_bbox().get_bottom() + 1);
+  RaycastResult result = Sector::get().get_first_line_intersection(eye,
+                                                                   Vector(eye.x, eye.y + DROP_DETECT_RANGE),
+                                                                   true,
+                                                                   nullptr);
+
+  if (!result.is_valid)
+  {
+    float sectorheight = static_cast<float>(Sector::get().get_editor_size().height * 32);
+    if (sectorheight <= eye.y + DROP_DETECT_RANGE + 1.f)
+    {
+      // Out of bounds. Drop to the lowest point possible by faking
+      // a raycast result.
+      if (s_ground_height < 0.f)
+        s_ground_height = sectorheight;
+      result.box = Rectf(Vector(0.f, s_ground_height), Sizef(1.f, 1.f));
+    }
+    else
+    {
+      return false;
+    }
+  }
+
+  m_state = STATE_DROPPING;
+  m_target_height = std::max(result.box.get_top() - 16.f - get_bbox().get_height(), 0.f);
+  m_timer.start(calculate_time(DROP_TIME));
+
+  return true;
+}
+
+bool
+Tarantula::hang_to(float height, float time, bool calctime, State nextstate, EasingMode easing, const std::string& action)
+{
+  set_action(action, 0);
+
+  bool is_newstate = (m_target_height != height);
+  m_target_height = height;
+
+  if (!m_timer.started())
+  {
+    if (!is_newstate)
+    {
+      // Timer finished
+      m_state = nextstate;
+      m_last_height = height;
+      return true;
+    }
+    else
+    {
+      m_timer.start(calctime ? calculate_time(time) : time);
+    }
+  }
+
+  double progress = static_cast<double>(m_timer.get_timegone() / m_timer.get_period());
+  float offset = static_cast<float>(getEasingByName(easing)(progress)) * (m_target_height - m_last_height);
+
+  Vector pos(get_bbox().get_left(), m_last_height + offset);
+  set_pos(pos);
+
+  return false;
+}
+
+float
+Tarantula::calculate_time(float div)
+{
+  if (div <= 0.f)
+    return 0.f;
+
+  return std::abs(m_target_height - get_bbox().get_top()) / div;
+}
+
+void
+Tarantula::collision_solid(const CollisionHit& hit)
+{
+  BadGuy::collision_solid(hit);
+
+  if (hit.top)
+    m_attach_ceiling = true;
+
+  if (m_was_grabbed && m_frozen && hit.bottom)
+    kill_fall();
+}
+
+void
+Tarantula::draw(DrawingContext& context)
+{
+  BadGuy::draw(context);
+
+  if (BadGuy::get_state() == STATE_FALLING ||
+      BadGuy::get_state() == STATE_SQUISHED ||
+      Editor::is_active())
+    return;
+
+  //FIXME: Hello reviewers, is there a better way to center the silk horizontally? Thank you.
+  Vector pos(get_bbox().get_left() + ((get_bbox().get_width() - static_cast<float>(m_silk->get_width()))/2),
+             m_start_position.y - 32.f);
```

Comment:
I don't think so, however since Tarantula doesn't move horizontally you can cache silk X position somewhere.

Reply:
it does move horizontally if not static

Reply:
Ah alright